cachi2/core/package_managers/yarn/resolver.py

```diff
@@ -124,44 +141,217 @@ def resolve_packages(source_dir: RootedPath) -> list[Package]:
+            # workspace dependencies have reliable names but report '0.0.0-use.local' as the version
+            name = _scoped_name(locator)
+            _, version = self._read_name_version_from_packjson(packjson)
+        elif isinstance(locator, (FileLocator, HttpsLocator)):
+            if not package.cache_path:
+                raise _CouldNotResolve(
+                    "expected a zip archive in the cache but 'yarn info' says there is none",
+                )
+            cache_path = self._cache_path_as_rooted(package.cache_path)
+            if not cache_path.path.exists():
+                raise _CouldNotResolve(
+                    f"cache archive does not exist: {cache_path.subpath_from_root}"
+                )
+            log_for_locator("reading package name from %s", cache_path.subpath_from_root)
+            # file and https dependencies have reliable versions but unreliable names
+            name = self._read_name_from_cache(cache_path)
+            version = package.version
+        elif isinstance(locator, (PortalLocator, LinkLocator)):
+            parent_locator = locator.locator
+            packjson = self._project_subpath(
+                parent_locator.relpath, locator.relpath, "package.json"
+            )
+            if isinstance(locator, LinkLocator) and not packjson.path.exists():
+                # if a link dependency doesn't have a package.json, we have to rely on the locator
+                name = _scoped_name(locator)
+                version = None
+            else:
+                # otherwise, take both the name and the version from package.json
+                # (name is unreliable, version is '0.0.0-use.local')
+                log_for_locator(
+                    "reading package name and version from %s", packjson.subpath_from_root
+                )
+                name, version = self._read_name_version_from_packjson(packjson)
+        elif isinstance(locator, PatchLocator):
+            if (
+                package.cache_path
+                # yarn info seems to always report the cache path for patch dependencies,
+                # but the path doesn't always exist
+                and (cache_path := self._cache_path_as_rooted(package.cache_path)).path.exists()
+            ):
+                log_for_locator("reading package name from %s", cache_path.subpath_from_root)
+                name = self._read_name_from_cache(cache_path)
+            elif orig_package := self._package_mapping.get(locator.package):
+                log_for_locator("resolving the name of the original package")
+                name = self._resolve_package(orig_package).name
+            else:
+                raise _CouldNotResolve(
+                    "the 'yarn info' output does not include either an existing zip archive "
+                    "or the original unpatched package",
+                )
+            version = package.version
+        else:
+            assert_never(locator)
```

Comment:
My concern with asserts in non-test code is *always* "The current code generator emits no code for an assert statement when optimization is requested at compile time" (e.g. command line option [-O]) - asserts are simply remove.
*We* probably wouldn't run it that way, but someone else might, and years from now, when none of us are actively working on cachi2 any more, someone inside RH mght even think it's a good idea to run with `-O`...

Reply:
It's not really an assert, it's a type-checker trick to check that this branch can never be reached at runtime: https://typing.readthedocs.io/en/latest/source/unreachable.html#assert-never-and-exhaustiveness-checking

If the `mypy` check is passing, then this code is unreachable. If cachi2 somehow does reach this branch, it will fail with a `NameError: name 'assert_never' is not defined`, because we only import it `if TYPE_CHECKING`

In python 3.11 and later, you can import `assert_never` directly from `typing` and don't have to import it conditionally from `typing_extensions`. Then it would fail with an `AssertionError: Expected code to be unreachable, but got: ...` regardless of any optimizations (the function doesn't use `assert`, it raises the error directly)

Reply:
😵‍💫

Reply:
Weird. Using a *type* checker for that feels like [crossing the streams](https://quotegeek.com/quotes-from-movies/ghostbusters/206/). OTOH, I wouldn't know how else to automagically check something like that...

tl;dr: TIL :grin:

Reply:
If you inline the function, it becomes:

```python
if isinstance(locator, ...):
   ...
else:
    locator: Never = locator
    raise AssertionError(f"Expected code to be unreachable, but got: {locator!r}")
```

Remove the line with Never and you're not using the type checker, you're raising an exception as usual. Keep the line and you're doing the same thing + you can statically check that you covered all possible cases

Reply:
It gets a bit weirder if you remember that we're importing assert_never conditionally. Maybe it would help to inline the check instead of using assert_never? Or re-implement assert_never as a util and import it from our own code?

Reply:
Added